### DISCUSSION TITLE
Client Timeout API

### DISCUSSION THREAD

**Discussion by fantix** [geldata/gel#3352] [Pre-RFC]
This is not about protocol and server-side timeouts, please check #1852 for that.

Based on edgedb-python, now we have the following timeouts on the client side:

1. `wait_until_available`: when a new connection is needed, the total time to wait until a new connection is established. The client will keep reconnecting until established or timed out.
2. (Connect) `timeout`: the time to wait for a single connect action. Note that reconnect may happen within a query or transaction depending on the retry options.
3. Client close timeout: the client will wait up to this time for all connections to be returned to the pool and closed gracefully, or else it'll brutally terminate any lingering connections.

`wait_until_available` and `timeout` are connect parameters in `create_*client()`, and `asyncio.wait_for()` is needed to set a timeout on closing a Python asyncio client.

When it comes to query timeouts, we could have different APIs:

1. Managed by framework like using `asyncio.wait_for()` or `Task.cancel()`.
2. Providing an additional `timeout` query parameter.
3. Use the options mixin like `client.with_timeout(command_timeout=5).query(...)`.
4. Use the connect parameter like `create_client(command_timeout=5)`.
5. Use a custom context like `with client.timeout(5):`.

Before #1852 is implemented, the client shall raise a `ClientTimeoutError` on timeout except that (1) would raise `asyncio.TimeoutError` or `asyncio.CancelledError`. The connection shall be brutally disconnected, expecting the server to cancel the undergoing query in Postgres if any.

* (1) should be supported anyways for Python asyncio client.
* (5) could leverage thread-locals or PEP-567 context variables.
* (3) and (4) could work together.
* It would be great if different language bindings share similar conventions.

**Comment by tailhook** [comment]
I think (3) is right choice given the overall shape of our API.

Additional (4) creates connection-time parameters, which we want to avoid. And adds another way to set parameters. Some users may assume that there is only a timeout on connection and will never learn (3). While (3) as a single API allows both use cases.

**Reply by elprans** [reply]
Yeah, it seems that way for me as well.  It'd be weird to have `.with_retry_options()`, but not `.with_timeouts()` if we are talking about a per-query timeout.

**Reply by fantix** [reply]
Then perhaps we could use (5) to provide more clear semantics like this:

```python
with client.timeout(5):
    await client.query("SELECT 42")
    await client.query("SELECT 'foo'")
```

This means anything in the `with` block - including acquiring from the pool, reconnecting and waiting for server to be available, parsing and executing - should be done within 5 seconds.

Comparing to separate APIs like:

```python
client = edgedb.create_async_client(wait_until_available=30, timeout=10)
client_wt = client.with_timeout(command_timeout=5)
await client_wt.query("SELECT 42")
await client_wt.query("SELECT 'foo'")
```

The theroretical max time for *each* `client_wt.query()` call is 45 seconds (30 + 10 + 5), and it's ambiguous whether the 5 seconds is a shared time across all `client_wt` queires, or each query has its own 5-seconds timer.

**Reply by tailhook** [reply]
It's query timeout. Surely it's not shared between queries.

What you have written is achieved by `asyncio.wait_for` (or asyncio=timeout package for a nicer syntax). So fundamental block for asyncio, and doesn't have to be implemented.

Query/command timeout on the other hand, enforces thing on server, right? So it's unfortunate to send "remaining time" as timeout before each request request.

**Reply by elprans** [reply]
Yeah, a context manager implies that the specified timeout applies to the block _as a whole_ and not to each individual query in it.  Controlling per-block timeouts is better left to a language or third-party facility.

> Query/command timeout on the other hand, enforces thing on server, right?

Well, ideally the implementation would send the desired timeout in a message header and the server would then compile the query into something like `SET LOCAL statement_timeout = <timeout>; <your_query>;`.  This should be the most efficient way of _cancelling_ a slow query as well.  

However, this approach only takes into account query run time and ignores time spent waiting for the network (both between EdgeDB/Postgres and between EdgeDB/client).  This highlights an important point that there are actually two kinds of timeouts with different effect on resource usage and with different cancellation:

Statement timeouts guard against excessive CPU/RAM/Disk utilization on Postgres hosts and should ideally be cancelled by Postgres itself using `statement_timeout`.

Network timeouts between EdgeDB and clients are mostly a QA/HA issue, _except_ when in a transaction, but the latter should already be covered by the idle transaction timeout.

Network timeouts between EdgeDB and Postgres are trickier, because intermittent network lag will cause a Postgres connection to be busy reducing availability for other queries, but also because it's hard to distinguish a slow query from a network issue when you are waiting for a response to PostgreSQL `Execute`.

I think we need two different timeout knobs: "statement_execution_timeout" and "overall_timeout", where the latter should be greater than the former.  "statement_execution_timeout" gets sent to Postgres, but should _also_ be enforced by EdgeDB server to handle the EdgeDB/Postgres network issue.  "overall_timeout" is handled by the client the usual way.

We also need to flesh out if we need a separate retry policy for timeouts or not.

**Comment by elprans** [comment]
`connect` and `close` timeouts should be arguments, at least in the blocking client.

**Reply by fantix** [reply]
They are arguments in edgedb/edgedb-python#286, though the `connect` timeout argument is not immediately starting to count down on the call it's set: `create_client(timeout=8)` returns immediately, while `client.close(timeout=8)` will return within 8 seconds regardless of the graceful termination. This is okay and compatible, but the user needs to do the math like the example I added in [the previous reply](https://github.com/edgedb/edgedb/discussions/3352#discussioncomment-2008894).

But I agree `close` is different and should have an explicit `timeout` argument.

**Comment by 1st1** [comment]
I think we should go with:

> Use the options mixin like client.with_timeout(command_timeout=5).query(...).

> connect and close timeouts should be arguments, at least in the blocking client.
